Handle InternalError raised by cryptography when running in FIPS mode

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Closes #65459

There is a bug in the logic in `cryptography` that assumes the presence of an algorithm based on the `openssl` version.

Research from @wenottingham:

>TL;DR: the mere presence of paramiko in the ansible venv causes ansible to explode on startup, due to:

>ansible: see if paramiko is available if we need it
>ansible: import paramiko
>paramiko: check for key algorithms
>paramiko: is x25519 available?
>paramiko: let's generate a key to find out
>cryptography: is x25519 available?
>cryptography: well, your openssl is new enough, so I guess so
>cryptography: ok, generating a key
>openssl: HA HA HA DISABLED BY FIPS kaboom


##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/module_utils/compat/paramiko.py`
##### ADDITIONAL INFORMATION
I'm not sure about how to add tests for this.

samdoran committed:
Handle InternalError raised by cryptography when running in FIPS mode

sivel commented on lib/ansible/module_utils/compat/paramiko.py:
```diff
@@ -5,10 +5,15 @@
 from __future__ import absolute_import, division, print_function
 __metaclass__ = type
 
+from cryptography.exceptions import InternalError
```
Since cryptography isn't a dep for older paramiko versions, and we don't have a specific version requirement, we won't be able to use this.

sivel commented on lib/ansible/module_utils/compat/paramiko.py:
```diff
@@ -5,10 +5,15 @@
+# paramiko and gssapi are incompatible and raise AttributeError not ImportError
+# When running in FIPS mode, cryptography raises InternalError
+# https://bugzilla.redhat.com/show_bug.cgi?id=1778939
+except (ImportError, AttributeError, InternalError) as err:  
```
As discussed between me, @jborean93, and @abadger, let's just make this `except Exception as err:`

samdoran committed:
Catch Exception

Since older versions of paramiko do not require cryptography,
we cannot catch the exact exception from cryptography.

sivel commented on lib/ansible/module_utils/compat/paramiko.py:
```diff
@@ -10,5 +10,8 @@
+# paramiko and gssapi are incompatible and raise AttributeError not ImportError
+# When running in FIPS mode, cryptography raises InternalError
+# https://bugzilla.redhat.com/show_bug.cgi?id=1778939
+except (ImportError, AttributeError, Exception) as err:
```
Just nuke the other exceptions, and only leave `Exception`. You can keep documentation of the individual exceptions known things raise.

samdoran replied:
😊

samdoran committed:
Remove other exceptions since we're catching everything now

ryanpetrello commented:
I reproduced https://github.com/ansible/ansible/pull/65459 using Ansible Tower on RHEL 8.1 w/ FIPS enabled and saw:

![image](https://user-images.githubusercontent.com/214912/70349857-97f47300-1833-11ea-8186-c995f33833b1.png)

Next, I manually applied this patch to my ansible install and re-ran the same job:

![image](https://user-images.githubusercontent.com/214912/70349869-a2af0800-1833-11ea-840d-88d8b3f83578.png)
